Add Vision Transformer and ViTFeatureExtractor

NielsRogge opened the pull request:
# What does this PR do?

Opening up a new PR based on #10513 which uses @sgugger's new `image_utils.py` instead of `torchvision` for the image transformations, and is up-to-date with master.

Things to do:

- [x] fix one integration test (currently `ViTFeatureExtractor` converts the numpy arrays into DoubleTensors, but the model expects FloatTensors)
- [x] fix styling (`make style` is not working as expected on my machine, see remaining comments in previous PR)
- [x] perhaps change pooler logic? Design (and updated conversion script) currently at branch "add_pooler_to_vit"

cc @LysandreJik

sgugger commented on src/transformers/__init__.py:
```diff
@@ -1271,6 +1281,9 @@
         name for name in dir(dummy_flax_objects) if not name.startswith("_")
     ]
 
+# Vision-backed objects
+if is_vision_available():
+    _import_structure["models.vit"].append("ViTFeatureExtractor")
```
Can we put it in the same test as the `ImageFeatureExtractionMixin`?

NielsRogge replied:
What do you mean by "putting in the same test"?

sgugger replied:
Not writing a new `if is_vision_available()` test since there is already one before the modeling imports (line 347 on master)

sgugger commented on src/transformers/__init__.py:
```diff
@@ -2327,6 +2348,11 @@
         # They will raise an import error if the user tries to instantiate / use them.
         from .utils.dummy_flax_objects import *
+
+    # Vision-backed objects
+    if is_vision_available():
+        from .models.vit import ViTFeatureExtractor
```
Same here.

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+    for i in range(config.num_hidden_layers):
+        # encoder layers: output projection, 2 feedforward neural networks and 2 layernorms
+        rename_keys.append(
+            ("blocks." + str(i) + ".norm1.weight", "vit.encoder.layer." + str(i) + ".layernorm_before.weight")
```
```suggestion
            (f"blocks.{i}.norm1.weight", f"vit.encoder.layer.{i}.layernorm_before.weight")
```
A f-string is more readable here. Use the same for all rename_key down there.

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+            ("blocks." + str(i) + ".norm1.bias", "vit.encoder.layer." + str(i) + ".layernorm_before.bias")
+        )
+        rename_keys.append(
+            (
+                "blocks." + str(i) + ".attn.proj.weight",
+                "vit.encoder.layer." + str(i) + ".attention.output.dense.weight",
+            )
```
This should fit back in one line after using a f-string.

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+def read_in_q_k_v(state_dict, config, base_model=False):
+    for i in range(config.num_hidden_layers):
+        # read in weights + bias of input projection layer (in timm, this is a single matrix + bias)
+        in_proj_weight = state_dict.pop("blocks." + str(i) + ".attn.qkv.weight")
```
Same here and in the lines below, a f-string would be more readable.

sgugger replied:
Please do not resolve a comment without treating it. This line and the line below still need to use f-strings.

NielsRogge replied:
Ok, sorry I've missed these two

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+
+
+def remove_classification_head_(state_dict):
+    ignore_keys = [
+        "norm.weight",
+        "norm.bias",
+        "head.weight",
+        "head.bias",
+    ]
```
```suggestion
    ignore_keys = ["norm.weight", "norm.bias", "head.weight", "head.bias"]
```

sgugger commented on src/transformers/models/vit/convert_vit_timm_to_pytorch.py:
```diff
@@ -0,0 +1,268 @@
+
+import torch
+from PIL import Image
+from torchvision.transforms import Compose, Normalize, Resize, ToTensor
```
Can we use the `ImageFeatureExtractorMixin` as a standalone to do those?

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+]
+
+
+# Inspired by
+# https://github.com/rwightman/pytorch-image-models/blob/b9bd960a032c75ca6b808ddeed76bee5f3ed4972/timm/models/layers/helpers.py
+# From PyTorch internals
+def _ntuple(n):
+    def parse(x):
+        if isinstance(x, collections.abc.Iterable):
+            return x
+        return tuple(repeat(x, n))
+
+    return parse
+
+
+to_2tuple = _ntuple(2)
```
We don't need the n-version and just use the 2-version, so let's just write
```
def to_2tuple(x):
    if isinstance(x, collections.abc.Iterable):
        return x
    return (x, x)
```
instead of a general function we don't need.

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+        super().__init__()
+        if config.hidden_size % config.num_attention_heads != 0 and not hasattr(config, "embedding_size"):
+            raise ValueError(
+                "The hidden size (%d) is not a multiple of the number of attention "
+                "heads (%d)" % (config.hidden_size, config.num_attention_heads)
```
```suggestion
                f"The hidden size {config.hidden_size,} is not a multiple of the number of attention "
                f"heads {config.num_attention_heads}."
```
f-strings are better than format.

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+        self.attention.all_head_size = self.attention.attention_head_size * self.attention.num_attention_heads
+        self.pruned_heads = self.pruned_heads.union(heads)
+
+    def forward(
+        self,
+        hidden_states,
+        head_mask=None,
+        output_attentions=False,
+    ):
```
```suggestion
    def forward(self, hidden_states, head_mask=None,  output_attentions=False):
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+        head_mask=None,
+        output_attentions=False,
+    ):
+        self_outputs = self.attention(
+            hidden_states,
+            head_mask,
+            output_attentions,
+        )
```
```suggestion
        self_outputs = self.attention(hidden_states, head_mask, output_attentions)
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+        self.layernorm_before = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+        self.layernorm_after = nn.LayerNorm(config.hidden_size, eps=config.layer_norm_eps)
+
+    def forward(
+        self,
+        hidden_states,
+        head_mask=None,
+        output_attentions=False,
+    ):
```
```suggestion
    def forward(self, hidden_states, head_mask=None, output_attentions=False):
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+                    layer_head_mask,
+                )
+            else:
+                layer_outputs = layer_module(
+                    hidden_states,
+                    layer_head_mask,
+                    output_attentions,
+                )
```
```suggestion
                layer_outputs = layer_module(hidden_states, layer_head_mask, output_attentions)
```

sgugger commented on src/transformers/models/vit/modeling_vit.py:
```diff
@@ -0,0 +1,652 @@
+        # and head_mask is converted to shape [num_hidden_layers x batch x num_heads x seq_length x seq_length]
+        head_mask = self.get_head_mask(head_mask, self.config.num_hidden_layers)
+
+        embedding_output = self.embeddings(
+            pixel_values,
+        )
```
```suggestion
        embedding_output = self.embeddings(pixel_values)
```

sgugger commented on tests/test_feature_extraction_vit.py:
```diff
@@ -0,0 +1,219 @@
+        assert not (numpify and torchify), "You cannot specify both numpy and PyTorch tensors at the same time"
+
+        if equal_resolution:
+            image_inputs = [
+                np.random.randint(
+                    255, size=(self.num_channels, self.max_resolution, self.max_resolution), dtype=np.uint8
+                )
+                for i in range(self.batch_size)
+            ]
```
I'd make a for loop here too as the list comprehension with formatting is not super readable.

sgugger commented on tests/test_modeling_vit.py:
```diff
@@ -0,0 +1,367 @@
+def prepare_img():
+    url = "http://images.cocodataset.org/val2017/000000039769.jpg"
+    img = Image.open(requests.get(url, stream=True).raw)
+    return img
```
I think we should put this in the fixtures folder to avoid errors when connection is bad on the CI.

sgugger reviewed (approved):
Thanks for opening a clean PR. I have some more styling comments but we should be good to merge soon :-)

NielsRogge commented:
I've addressed all comments. The pooler now more closely matches the one of `BertModel`. 

Only `make fix-copies` is giving an error on CircleCI for now. Other than that the PR is ready.

sgugger commented on README.md:
```diff
@@ -238,6 +238,7 @@ Min, Patrick Lewis, Ledell Wu, Sergey Edunov, Danqi Chen, and Wen-tau Yih.
 1. **[XLM-RoBERTa](https://huggingface.co/transformers/model_doc/xlmroberta.html)** (from Facebook AI), released together with the paper [Unsupervised Cross-lingual Representation Learning at Scale](https://arxiv.org/abs/1911.02116) by Alexis Conneau*, Kartikay Khandelwal*, Naman Goyal, Vishrav Chaudhary, Guillaume Wenzek, Francisco Guzmán, Edouard Grave, Myle Ott, Luke Zettlemoyer and Veselin St ...
 1. **[XLNet](https://huggingface.co/transformers/model_doc/xlnet.html)** (from Google/CMU) released with the paper [​XLNet: Generalized Autoregressive Pretraining for Language Understanding](https://arxiv.org/abs/1906.08237) by Zhilin Yang*, Zihang Dai*, Yiming Yang, Jaime Carbonell, Ruslan Salakhutdinov, Quoc V. Le.
 1. **[XLSR-Wav2Vec2](https://huggingface.co/transformers/model_doc/xlsr_wav2vec2.html)** (from Facebook AI) released with the paper [Unsupervised Cross-Lingual Representation Learning For Speech Recognition](https://arxiv.org/abs/2006.13979) by Alexis Conneau, Alexei Baevski, Ronan Collobert, Abdelrahman Mohamed, Michael Auli.
+1. **[Vision Transformer (ViT)](https://huggingface.co/transformers/model_doc/vit.html)** (from Google AI) released with the paper [An Image is Worth 16x16 Words: Transformers for Image Recognition at Scale](https://arxiv.org/abs/2010.11929) by Alexey Dosovitskiy, Lucas Beyer, Alexander Kolesnikov, Dirk Weissenborn, Xiaohua Zhai, Thomas Unterthiner, Mostafa Dehghani, Matthias Minderer, Georg Heig ...
```
This should be put in the right place in alphabetical order. (Then rerun `make fix-copies` to fix the index too)

sgugger commented on tests/test_image_utils.py:
```diff
@@ -264,7 +264,9 @@ def test_normalize_image(self):
         # During the conversion rescale and channel first will be applied.
         expected = array.transpose(2, 0, 1).astype(np.float32) / 255.0
-        expected = (expected - np.array(mean)[:, None, None]) / np.array(std)[:, None, None]
+        expected = (expected - np.array(mean).astype(np.float32)[:, None, None]) / np.array(std).astype(np.float32)[
+            :, None, None
+        ]
```
This is starting to be a less unreadable. Can we stored `np_mean = np.array(mean).astype(np.float32)[:, None, None])` and `np_std = np.array(std).astype(np.float32)[:, None, None]` on the lines above?

sgugger reviewed (approved):
LGTM, Just a few final remarks before we merge!

LysandreJik reviewed (approved):
LGTM! Thanks for the implementation @NielsRogge!

sgugger commented:
Thanks for all your work on this @NielsRogge !
